make functions for new approaches to null fit based on constraint function

svteichman opened the pull request:
**Address issue #132** 

This PR does the following:
- updates the version to v3.0.0.0
- adds to the news.md file (we will want to update this eventually when this update is done) 
- in `emuFit()`, for each row of $\beta$ `k` that is being tested, figure out which option the constraint for row `k` is (symmetric, single category constraint, or other). I discussed this a bit with @adw96 , and we decided that for now for symmetric we'll just check whether the constraint function is the mean or a pseudo-Huber median, and if it is neither of those or a single category constraint then we set it to "other". This may miss other user-defined symmetric constraint functions, but we can always put a note somewhere in the code that if a user wants their symmetric constraint to go through option 1, then can submit a PR redirecting it that way. Or, another option is to add another argument to `emuFit()` aimed towards developers to bypass our current approach for determining the constraint option and to instead set it directly. I'll probably need to eventually implement this additional argument for `fastEmu`! 
- add a new argument to `score_test()` to take in this constraint option
- in `score_test()`, use the constraint option argument to either call `fit_null()` (the current fitting algorithm), `fit_null_scc()` (the single category constraint option) or `fit_null_symmetric()` (the mean or pseudo-Huber median option)
- create functions `fut_null_scc()` and `fit_null_symmetric()` that take in the same arguments as `fit_null()` and return the same values. Right now these two new functions simply call `fit_null()` internally as a placeholder to make sure everything runs without errors

I've tested locally that each of these types of constraints routes through the correct function. However, I haven't added any automated tests of these because I think that we'll want to do extensive testing, and I don't want to add tests now that will become redundant later and waste time when running the R-CMD-check. 

@ailurophilia, once you've populated these functions, let me know if you'd like me to add automated tests for these new functions and/or benchmark these new functions on data analyses. I'm super happy to help out with anything that would be useful, and am excited to see what you have in mind for these functions! 

Also both @ailurophilia and @adw96, let me know if there's anything you want me to add or modify to make this closer to your vision for this!

adw96 commented:
This is just wonderful -- great work and thank you, @svteichman !!! We both really appreciate it. 

Over to you, @ailurophilia !

ailurophilia commented:
Thank you both! I have some time carved out for this tomorrow. Not sure I can contribute to this PR, so think I'll populate the new null fitting functions you've made, @svteichman – thank you so much for that! – in a separate PR. If that will create a huge headache, though, let me know, and I'll adjust course!

adw96 commented:
Correct! Our process is for PRs to go through review, so I'm limiting push access where possible to avoid breaking changes. 

Looks like the correct workflow here is indeed to submit a new PR -- based on a quick glance at [this](https://stackoverflow.com/questions/22237609/adding-to-someone-else-s-pull-request-on-a-github-project-that-i-do-not-own). 

My apologies for suggesting a different workflow initially!!

adw96 commented:
Wooohoooo! @svteichman -- @ailurophilia just created #144 implementing the guts of this 🥳 how exciting!! 

When you have the bandwidth, would you mind going through #144, and integrating #144 and this PR in the way that you think is best? I think we can be relaxed about copying and pasting code rather than doing cherry-pick or patch or whatever git black magic is best practice here. 

Feel free to leave the Fisher scoring / analytical gradients for now (unless you want to take a crack at them!). 

Of course feel free to post questions or reach out offline. 

Thank you -- and if there's anything you want me to look at in advance of Wednesday, let me know and I'll do my best!!!

svteichman commented:
Ok will do! I'll plan to copy & paste the relevant parts of PR #144 into this one. For now, I'll include a flag for `fit_null_symmetric()` that reroutes it to the current null fitting algorithm by default, unless you want the code in #144 to be default for users in this version. Let me know which you prefer @adw96

svteichman commented:
@adw96 I took the updates from PR #144 and added them into this PR. This mostly involved copy & pasting, although I made two small changes: in `fit_null_symmetric()` I set `use_optim = TRUE` (which we can update once we have fisher scoring) and I removed the use of `tolerance` and replaced it with `B_tol` to be consistent with the `fit_null()` function, and updated the test in `test-fit_null_symmetric` accordingly  (@ailurophilia letting you know about these small updates). 

Let me know if there's anything I can do to help, or if there's a different way that you wanted these PRs reconciled.

adw96 commented:
Wow, that was so fast, @svteichman !! Thank you!!! Wow!!! Your efficiency always astounds me 🌟 

@svteichman How hard would it be to benchmark the time of this branch vs the current `main`? Can we also profile it to see how much time is spent doing the `optim` (which @ailurophilia conjectures is the current bottleneck slow)?

@ailurophilia if you do any more work on this, could you make edits from this branch? I think `gh pr checkout 133` then creating a new PR would be preferable to committing on the other PR. 

@svteichman should we delete the other PR? (pending confirmation from @ailurophilia )

svteichman commented:
I'll do some benchmarking to compare these two null fitting algorithms time-wise, and to see how much time optim is taking in the new approach. I think that the other PR could be deleted, and that the best approach would be the one you suggested, of @ailurophilia pulling from this PR and creating a new one.

svteichman commented:
Note from @adw96 on newest commits (implementing fisher scoring for new null fitting approach):

"This branch is related to https://github.com/statdivlab/radEmu/pull/133 . It takes the optim implementation of https://github.com/statdivlab/radEmu/pull/144 and adds a Fisher scoring approach. Definitely standing on algorithm and coding giants.

My initial tests -- which were not extensive -- suggested this could reduce speed by 20-800% -- pretty amazing!!

This is still a WIP, with a couple of issues remaining.

I have changed the default of fit_null_symmetric to route to Fisher scoring. test-emuFit.R:380:5 now fails "emuFit runs with just intercept model." I presume its the sandwich approach, but will ask @svteichman to investigate.
I'm still investigating test-emuFit.R:434.
I think I'm not following conventions for how we fail gracefully. I have a stop(...) call in fit_null_symmetric.
@svteichman could I ask your help implementing the appropriate try catches? (I have to move on from this soon, unfortunately...)
Ditto for what's happening in "remove_zero_comparison_pvals argument works in different ways", test-zero_comparison_check.R:72.
@svteichman Feel free to begin your review -- I appreciate your questions and suggestions! I also welcome your benchmarking."

svteichman commented:
With the above commit, all tests pass for me locally. The major change I made is that when `fit_null_symmetric()` fails `ntries` times, I then run `fit_null` to try to fit. Even if this fails, unlike `fit_null_symmetric()` it should (as I can tell) always return something, avoiding the previous error. I also think this is a reasonable approach (if you can't fit in our new and improved way, then try the other null fitting algorithm). For the intercept only model, I updated to make sure dimensions match when needed for `p = 1`.

svteichman commented:
Update on this - the toy examples where `fit_null_symmetric()` is failing were coming from a bug in how the info matrix was constructed in `null_repar_ll_gr()` for `p` > 2. That is fixed in the following commit, and all toy examples converge. In small benchmarking tests, the new null is a similar speed or slower for `J` = 10, and typically about 10x faster for `J` >= 50 (with some trials in which the new null is slower as well). A larger benchmarking simulation to come (checking both speed and similarity of resulting score test p-values).

svteichman commented on R/fit_null_scc.R:
```diff
@@ -0,0 +1,97 @@
+                         trackB = FALSE 
+) {
+  
+  ### David to replace - for now redirecting to fit_null so that everything still runs! 
```
The goal of this was to have an option to eventually put in a simpler algorithm for fitting a single category constraint, but we haven't worked on this yet so this is just a placeholder to route through this. So I think it could stay but for now it doesn't do anything. The comment can be removed though - this isn't for David to replace, that's just what I had thought for some reason when I put this here a few months back.

svteichman commented on R/get_constrained_gr.R:
```diff
@@ -0,0 +1,96 @@
+    return(gr)
+  }
+  # 
+  # r_info <- restr_info(log_means_js,js,j_constr,k_constr,j_ref,J)
```
I don't think that I put this here or commented it out, but I believe that it should be removed because it is no longer relevant to the code.
